### ISSUE TITLE
Add "strict dtype checking" flag

### ISSUE THREAD

**Issue by axch** [google/jax#4966]
Please add a configuration flag to allow strict dtype checking, namely never implicitly casting dtypes but always raising an error when mixing, say, float32 and float64 (like TensorFlow's dtype semantics, for example).

This would be very useful for debugging dtype mistakes, which in turn is very useful for debugging floating-point accuracy by running the same computation in float32 and float64.

**Comment by axch** [comment]
@jakevdp ?

**Comment by jakevdp** [comment]
I doubt we'll add other global configurations to JAX. Such modalism is problematic for many reasons.

Note that `jax.lax` behaves in the way that you desire: it does no implicit type promotions, and requires all operations to have strictly matching types. If you prefer that mode of working, you might try expressing your operations that way.

**Comment by axch** [comment]
Fair. Here's the user story from which this request arose: A colleague developed some hairy Bayesian analysis in a colab. They have evidence that running on TPUs is breaking their results because of numerical accuracy. However, they want to run on TPUs for the speed. So they come to me, hoping I can find places to improve numerics in their code. Now I have maybe 700 lines of code I didn't write, and I'd like to run it in float32 and in float64 on a variety of inputs to see where it might be losing accuracy. Needless to say, the less rewriting of their code I have to do the better.

I can respect not wanting to add global flags (though I suggested one because Jax already has one, without knowing that you are trying to get rid of them). Could this be better done with a transformation?

**Comment by SamPruden** [comment]
I can complement with a far more basic user story.

I have a bad habit of typing integer literals when I should be typing float literals, and that causes some silly little problems while iterating on my model design. I've messed up benchmarks by accidentally testing on integers instead of floats, for example. I've taken to using `assert` statements to catch these things, but it's a bit of unnecessary clutter. These mistakes are my own fault, but it would improve the development experience if JAX could fail or at least warn when these things happen implicitly during a debug workflow.

@axch provides a far more compelling argument, but I believe this could be a nice quality of life thing for many users.

**Comment by SamPruden** [comment]
I currently use `config.update("jax_debug_nans", True)` while developing, and a similar `config.update("jax_debug_implicit_casts", True)` would be a nice complement for that.

**Comment by mattjj** [comment]
> Could this be better done with a transformation?

I think this couldn't quite be done with a transformation because dtype promotion happens "above" the primitive level; the primitives in `jax.lax` don't do automatic dtype promotion (as Jake already said), but instead [most dtype promotion happens in lax_numpy.py, in the `_promote_dtypes` function](https://github.com/google/jax/blob/cc2ec6143dac4ecfd667cd1b49990232e399415b/jax/_src/numpy/lax_numpy.py#L267). (Funny how the primitives in `lax` are decidedly not _lax_ about types...)

We had a similar ask for more strictness (less promotion) in the context of [`jax.numpy` automatic rank promotion](https://jax.readthedocs.io/en/latest/rank_promotion_warning.html), and just above the previously linked `_promote_dtypes` function in lax_numpy.py you can see [the implementation of disable-able rank promotion](https://github.com/google/jax/blob/cc2ec6143dac4ecfd667cd1b49990232e399415b/jax/_src/numpy/lax_numpy.py#L237-L265). I think we could do something similar for dtypes in a similar number of lines of code (i.e. O(10)). So if we were ever to consider adding another global flag, I think adding one here would be defensible, though it's not totally clear whether we should.

Happily, instead of waiting for us to decide on and implement a change that might affect a lot of users, as a workaround you can just monkey-patch that function yourself! That is, you can write something like

```python
import jax._src.numpy.lax_numpy as lnp
from jax import lax

def _dont_promote_dtypes(*args):
  if len(args) < 2:
    return args
  else:
    to_dtype = lnp.result_type(*args)
    if any(to_dtype != lnp.result_type(x) for x in args):
      raise Exception
    return [lax.convert_element_type(x, to_dtype) for x in args]

lnp._promote_dtypes = _dont_promote_dtypes

###

import jax.numpy as jnp
jnp.add(1, 2.)
```

Not all NumPy functions call that dtype promotion wrapper, but quite a few do.

WDYT?

**Comment by SamPruden** [comment]
Perhaps that monkey-patching solution is worth mentioning in the [Common Gotchas](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html) documentation alongside a suggestion to monkey-patch to print a warning on promotion. That's where the instructions for debugging NaNs are. It's a good trick to have up your sleeve for tricky situations, but hard to discover oneself without familiarity with the source.

A more robust solution in the future would be welcome, but I don't have strong opinions on it.

**Comment by jakevdp** [comment]
This has now been done (see https://github.com/google/jax/pull/10824)
